src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -0,0 +1,1305 @@
+
+        CheckInvariants();
+    }
+
+    public SolutionCompilationState(
+        SolutionState solution,
+        bool partialSemanticsEnabled)
+        : this(
+              solution,
+              partialSemanticsEnabled,
+              projectIdToTrackerMap: ImmutableDictionary<ProjectId, ICompilationTracker>.Empty,
+              frozenSourceGeneratedDocument: null)
+    {
+    }
+
+    public SolutionServices Services => this.Solution.Services;
+
+    private void CheckInvariants()
+    {
+        // Only run this in debug builds; even the .Any() call across all projects can be expensive when there's a lot of them.
+#if DEBUG
+        // An id shouldn't point at a tracker for a different project.
+        Contract.ThrowIfTrue(_projectIdToTrackerMap.Any(kvp => kvp.Key != kvp.Value.ProjectState.Id));
+#endif
+    }
+
+    public SourceGeneratedDocumentState? FrozenSourceGeneratedDocumentState => _frozenSourceGeneratedDocumentState;
+
+    private SolutionCompilationState Branch(
+        SolutionState newSolutionState,
+        ImmutableDictionary<ProjectId, ICompilationTracker>? projectIdToTrackerMap = null,
+        Optional<SourceGeneratedDocumentState?> frozenSourceGeneratedDocument = default)
+    {
+        projectIdToTrackerMap ??= _projectIdToTrackerMap;
+        var newFrozenSourceGeneratedDocumentState = frozenSourceGeneratedDocument.HasValue ? frozenSourceGeneratedDocument.Value : _frozenSourceGeneratedDocumentState;
+
+        if (newSolutionState == this.Solution &&
+            projectIdToTrackerMap == _projectIdToTrackerMap &&
+            newFrozenSourceGeneratedDocumentState == _frozenSourceGeneratedDocumentState)
+        {
+            return this;
+        }
+
+        return new SolutionCompilationState(
+            newSolutionState,
+            PartialSemanticsEnabled,
+            projectIdToTrackerMap,
+            newFrozenSourceGeneratedDocumentState);
+    }
+
+    /// <inheritdoc cref="SolutionState.ForkProject"/>
+    private SolutionCompilationState ForkProject(
+        (SolutionState newSolutionState, ProjectState oldProjectState, ProjectState newProjectState) stateTuple,
+        CompilationAndGeneratorDriverTranslationAction? translate,
+        //ProjectDependencyGraph? newDependencyGraph = null,
+        bool forkTracker)
+    {
+        return ForkProject(
+            stateTuple.newSolutionState,
+            stateTuple.newProjectState,
+            translate,
+            forkTracker);
+    }
+
+    /// <inheritdoc cref="SolutionState.ForkProject"/>
+    private SolutionCompilationState ForkProject(
+        SolutionState newSolutionState,
+        ProjectState newProjectState,
+        CompilationAndGeneratorDriverTranslationAction? translate,
+        //ProjectDependencyGraph? newDependencyGraph = null,
+        bool forkTracker)
+    {
+        // If the spsolution didn't actually change, there's no need to change us.
+        if (newSolutionState == this.Solution)
+            return this;
```

Comment:
this is the common check we want from almost *all* update methods.  If the method didn't go through and actually manipulate the solution-state in any way, then there's nothing to be done with the compilation state.

There is *one* exception (which i will call out below).  Which is why we have ForceForkProject as well.

Reply:
Maybe call this ForkProjectIfStateChanged just so that behavior is a bit more clear?